### PR TITLE
[RISCV] Support __builtin_cpu_init and __builtin_cpu_supports

### PR DESCRIPTION
**Pull request by preames**
This implements the __builtin_cpu_init and __builtin_cpu_supports builtin routines based on the compiler runtime changes in https://github.com/llvm/llvm-project/pull/85790.

This is inspired by https://github.com/llvm/llvm-project/pull/85786. Major changes are a) a restriction in scope to only the builtins (which have a much narrower user interface), and the avoidance of false generality.  This change deliberately only handles group 0 extensions (which happen to be all defined ones today), and avoids the tblgen changes from that review.

I don't have an environment in which I can actually test this, but @BeMg has been kind enough to report that this appears to work as expected.  

Before this can make it into a release, we need a change such as https://github.com/llvm/llvm-project/pull/99958.  The gcc docs claim that cpu_support can be called by "normal" code without calling the cpu_init routine because the init routine will have been called by a high priority constructor.  Our current compiler-rt mechanism does not do this.

### CODE REVIEW

File: llvm/lib/TargetParser/RISCVISAInfo.cpp [hunk 1]
```diff
@@ -1020,3 +1020,64 @@ std::string RISCVISAInfo::getTargetFeatureForExtension(StringRef Ext) {
   return isExperimentalExtension(Name) ? "experimental-" + Name.str()
                                        : Name.str();
 }
+
+struct RISCVExtBit {
+  const StringRef ext;
+  uint64_t bitpos;
```
Comment (BeMg) [inline llvm/lib/TargetParser/RISCVISAInfo.cpp:1026]: Does it need `uint64_t` for bit position? I think the value doesn't exceed 63.

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 2]
```diff
@@ -62,6 +62,8 @@
 #include "llvm/Support/MathExtras.h"
 #include "llvm/Support/ScopedPrinter.h"
 #include "llvm/TargetParser/AArch64TargetParser.h"
+#include "llvm/TargetParser/RISCVISAInfo.h"
+#include "llvm/TargetParser/RISCVTargetParser.h"
```
Comment (BeMg) [inline clang/lib/CodeGen/CGBuiltin.cpp:66]: `RISCVTargetParser.h` can be dropped due to its come from tablegen stuff.

**Comment by BeMg** [comment]
> * Updating clang tests to exercise this logic.
> * Actually running it at all.  I did not build compiler-rt, and thus all my checking was of generated asm/IR.

Maybe we could compile the test code with `compiler-rt/lib/builtins/riscv/feature_bits.c` manually. Then we could get rid of copmiler-rt building during running the `__builtin_cpu_init` and `__builtin_cpu_supports` actually. 

Actually, I tried this patch in my local, and it work well. 

---

Here is build script for compiler-rt, if you want to build compiler-rt.

```
cmake -DCMAKE_C_COMPILER_TARGET="riscv64-unknown-linux-gnu" \
      -DCMAKE_ASM_COMPILER_TARGET="riscv64-unknown-linux-gnu" \
      -DCMAKE_AR=./path/to/llvm/install/bin/llvm-ar \
      -DCMAKE_C_COMPILER=./path/to/llvm/install/bin/clang \
      -DCMAKE_NM=./path/to/llvm/install/bin/llvm-nm \
      -DCMAKE_RANLIB=./path/to/llvm/install/bin/llvm-ranlib \
      -DCOMPILER_RT_BUILD_BUILTINS=ON \
      -DCOMPILER_RT_BUILD_LIBFUZZER=OFF \
      -DCOMPILER_RT_BUILD_MEMPROF=OFF \
      -DCOMPILER_RT_BUILD_PROFILE=OFF \
      -DCOMPILER_RT_BUILD_SANITIZERS=OFF \
      -DCOMPILER_RT_BUILD_XRAY=OFF \
      -DCOMPILER_RT_DEFAULT_TARGET_ONLY=ON \
      -DLLVM_CONFIG_PATH=./path/to/llvm/install/bin/llvm-config \
      -G "Ninja" ../compiler-rt
ninja 
# Put the `libclang_rt.builtins-riscv64.a` in correct path
cp ./path/to/build/lib/linux/* ./path/to/llvm/install/lib/clang/19/lib/linux
```

```
# Compile using compiler-rt
clang -march=rv64imafd -rtlib=compiler-rt test.c
```

> * Investigate claims from gcc docs that __builtin_cpu_init is called early in process lifetime with high priority constructor.  I did not find this with some quick searching.

File: llvm/lib/TargetParser/RISCVISAInfo.cpp [hunk 3]
```diff
@@ -1020,3 +1020,64 @@ std::string RISCVISAInfo::getTargetFeatureForExtension(StringRef Ext) {
...
+};
+
+/// Maps extensions with assigned bit positions within group 0 of
+/// __riscv_features_bits to their respective bit position.  At the
+/// moment all extensions are within group 0.
+static RISCVExtBit RISCVGroup0BitPositions[] = {
```
Comment (BeMg) [inline llvm/lib/TargetParser/RISCVISAInfo.cpp:1032]: One reason we choose the `tablegen` approach is trying to maintain the everything relate to extension (In this case is groupID/bitPos) inside `RISCVFeatures.td`. 

How about land https://github.com/llvm/llvm-project/pull/94440 now, and replace this struct with tablegen generated `.inc` file? Or maybe we could replace it after 19.x release?

Comment (preames) [reply llvm/lib/TargetParser/RISCVISAInfo.cpp:1032]: IMO, adding the additional information to RISCVFeatures.td just complicates things.

File: llvm/lib/TargetParser/RISCVISAInfo.cpp [hunk 4]
```diff
@@ -1020,3 +1020,43 @@ std::string RISCVISAInfo::getTargetFeatureForExtension(StringRef Ext) {
   return isExperimentalExtension(Name) ? "experimental-" + Name.str()
                                        : Name.str();
 }
+
+struct RISCVExtBit {
+  const StringRef ext;
```
Comment (topperc) [inline llvm/lib/TargetParser/RISCVISAInfo.cpp:1025]: Use StringLiteral.

File: llvm/lib/TargetParser/RISCVISAInfo.cpp [hunk 5]
```diff
@@ -1020,3 +1020,43 @@ std::string RISCVISAInfo::getTargetFeatureForExtension(StringRef Ext) {
...
+  uint8_t bitpos;
+};
+
+/// Maps extensions with assigned bit positions within group 0 of
+/// __riscv_features_bits to their respective bit position.  At the
+/// moment all extensions are within group 0.
+static RISCVExtBit RISCVGroup0BitPositions[] = {
```
Comment (topperc) [inline llvm/lib/TargetParser/RISCVISAInfo.cpp:1032]: Make this `constexpr` to make sure the strlen for the strings actually happens at compile time.

File: clang/test/Preprocessor/has_builtin_cpuid.c [hunk 6]
```diff
@@ -2,14 +2,29 @@
...
 #   error "ARM/PPC shouldn't have __builtin_cpu_init"
 # endif
+#else
+# ifdef RISCV
+#   error "RISCV should have __builtin_cpu_init"
+# endif
+#endif
+
+#if !__has_builtin(__builtin_cpu_supports)
+# if defined(X86) || defined(RISCV)
```
Comment (topperc) [inline clang/test/Preprocessor/has_builtin_cpuid.c:27]: Does ARM support __builtin_cpu_supports?

**Review by BeMg — APPROVED** [review]
LGTM. Work fine with current compiler-rt `__init_riscv_feature_bits `/`__riscv_feature_bits ` and qemu with hwprobe support. (call `__builtin_cpu_init` manually situation)

**Comment by preames** [comment]
I've gone ahead and merged this into main.  We have missed the branch creation, so without further action this will not be included in 19.x.  We need to ensure the constructor change for compiler-rt lands, and then backport them together if we choose to.

**Pull request merged**